src/Data/README.md

```diff
@@ -0,0 +1,110 @@
+# Datatypes for ILIAS
+
+This service should contain standard datatypes for ILIAS that are used in many
+locations in the system and do not belong to a certain service.
+
+Other examples for data types that could (and maybe should) be added here:
+
+* Option (akin to rusts type)
+* (il)Datetime
+* ObjectId, ReferenceId
+* HTML, Text, Markdown
+* List<int>, List<bool>, ...
+
+This is not to be confused with the service for types. This services is about
+the data, not the types thereof. It still uses types to talk about data (like
+a lot of code does), but it does not reify types as data (which the service
+for types and the PHP-ReflectionClass does).
+
+## Result
+
+A result encapsulates a value or an error and simplifies the handling of those.
+
+```php
+<?php
+require_once(__DIR__."/Factory.php");
+
+$f = new Data/Factory;
+
+// Build a value that is ok.
+$pi = $f->ok(3.1416);
+
+// Value is ok and thus no error.
+assert($pi->isOK());
+assert(!$pi->isError());
+
+// Do some transformation with the value.
+$r = 10;
+$A = $pi->map(function($value_of_pi) use ($r) { return 2 * $value_of_pi * $r; });
+
+// Still ok and no error.
+assert($A->isOk());
+assert(!$A->isError());
+
+// Retrieve the contained value.
+$A_value = $A->value();
+assert($A_value == 2 * 3.1415 * 10);
+
+// No error contained...
+$raised = false;
+try {
+	$A->error();
+	assert(false); // Won't happen, error raises.
+}
+catch(\LogicException $e) {
+	$raised = true;
+}
+assert($raised);
+
+
+// Build a value that is not ok.
+$e = $f->error("There was some error...");
+
+// This is of course an error.
+assert(!$e->isOK());
+assert($e>isError());
+
+// Transformations are nops.
+$A = $f->map($function($v) { assert(false); });
+
+// Attempts to retrieve the value will throw.
+$raised = false;
+try {
+	$A->value()
+	assert(false); // Won't happen.	
+}
+catch (\ILIAS\Data\NotOKException $e) {
+	$raised = true;
+}
+assert($raised);
+
+// For retrieving a default could be supplied.
+$v = $e->valueOr("default");
```

Comment:
Yes, totally. Thx!

I'm thinking about how we could put snippets from READMEs under test, it might just be `require_once("README.md")` wrapped in a PHPUnit test.